pkg/services/notary/notary.go

```diff
@@ -0,0 +1,376 @@
+	Config struct {
+		MainCfg config.P2PNotary
+		Chain   blockchainer.Blockchainer
+		Log     *zap.Logger
+	}
+)
+
+// request represents Notary service request.
+type request struct {
+	typ RequestType
+	// isSent indicates whether main transaction was successfully sent to the network.
+	isSent bool
+	main   *transaction.Transaction
+	// minNotValidBefore is the minimum NVB value among fallbacks transactions.
+	// We stop trying to send mainTx to the network if the chain reaches minNotValidBefore height.
+	minNotValidBefore uint32
+	fallbacks         []*transaction.Transaction
+	// nSigs is the number of signatures to be collected.
+	// nSigs == nKeys for standard signature request;
+	// nSigs <= nKeys for multisignature request.
+	// nSigs is 0 when all received requests were invalid, so check request.typ before access to nSigs.
+	nSigs uint8
+	// nSigsCollected is the number of already collected signatures
+	nSigsCollected uint8
+
+	// sigs is a map of partial multisig invocation scripts [opcode.PUSHDATA1+64+signatureBytes] grouped by public keys
+	sigs map[*keys.PublicKey][]byte
+}
+
+// NewNotary returns new Notary module.
+func NewNotary(bc blockchainer.Blockchainer, log *zap.Logger) (*Notary, error) {
+	cfg := bc.GetConfig().P2PNotary
+	w := cfg.UnlockWallet
+	wallet, err := wallet.NewWalletFromFile(w.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	haveAccount := false
+	for _, acc := range wallet.Accounts {
+		if err := acc.Decrypt(w.Password); err == nil {
+			haveAccount = true
+			break
+		}
+	}
+	if !haveAccount {
+		return nil, errors.New("no wallet account could be unlocked")
+	}
+
+	return &Notary{
+		requests: make(map[util.Uint256]*request),
+		Config: Config{
+			MainCfg: cfg,
+			Chain:   bc,
+			Log:     log,
+		},
+		wallet: wallet,
+	}, nil
+}
+
+// OnNewRequest is a callback method which is called after new notary request is added to the notary request pool.
+func (n *Notary) OnNewRequest(payload *payload.P2PNotaryRequest) {
+	if n.getAccount() == nil {
+		return
+	}
+
+	nvbFallback := payload.FallbackTransaction.GetAttributes(transaction.NotValidBeforeT)[0].Value.(*transaction.NotValidBefore).Height
+	nKeys := payload.MainTransaction.GetAttributes(transaction.NotaryAssistedT)[0].Value.(*transaction.NotaryAssisted).NKeys
+	typ, nSigs, pubs, validationErr := n.verifyIncompleteWitnesses(payload.MainTransaction, nKeys)
+	n.reqMtx.Lock()
+	defer n.reqMtx.Unlock()
+	r, exists := n.requests[payload.MainTransaction.Hash()]
+	if exists {
+		for _, fb := range r.fallbacks {
+			if fb.Hash().Equals(payload.FallbackTransaction.Hash()) {
+				return // then we already have processed this request
+			}
+		}
+		if nvbFallback < r.minNotValidBefore {
+			r.minNotValidBefore = nvbFallback
+		}
+		if r.typ == Unknown && validationErr == nil {
+			r.typ = typ
+			r.nSigs = nSigs
+		}
+		if r.typ != Unknown && r.nSigsCollected >= r.nSigs { // already collected sufficient number of signatures to complete main transaction
```

Comment:
This can be moved down below, before `if validationErr == nil {` to not duplicate fallbacks appending.

Reply:
Fixed.